### ISSUE TITLE
topk selection

### ISSUE THREAD

**Issue by WilliamTambellini** [arrayfire/arrayfire#1870]
Here is a feature request that is more and more needed for MachineLearning/NeuralNetworks but should better be implemented in AF at lowlevel (cuda, ...).

Feature: find the top k values from an input array.
Example of expected behavior:
input [4 2 1 1]
    0.399     0.735
    0.672     0.516
    0.533     0.393
    0.138     0.270
 
topk(2) desired indices output [2 2 1 1]
         1          0
         2          1
outputing the top values would be nice to have also, just like af::sort() does.
 
The best CPU way to do it is probably calling .host() and then std::partial_sort() or std::nth_element().

I have tried some POC with the current version of AF using af::sort(), ie full sorting.
Sadly the CPU way is still faster than GPU (at least on GTX10X0) at least for such sizes : 
- k about 10 and for an input of about 100k rows. 
- 1d vector of 200k rows, k = 5

Here are the algos I ve tried with current version of AF:
- full sort (af::sort) and then filtering the sorted arrays using af::seq : still slower than CPU
- loop of af::max() : looping k times, extracting the max values and invalidating the max value before iterating. Pretty fast for very small values of k (2, 3, ...) but then slower than CPU for higher value. 

Some refs:
- pytorch topk:  void THTensor_(topk)(THTensor *rt_, THLongTensor *ri_, THTensor *t, long k, int dim, int dir, int sorted)
https://github.com/pytorch/pytorch/blob/master/torch/lib/TH/generic/THTensorMath.c#L2547-L2566
- cutorch radixSelect: (source joined to this ticket) using radixSelect() in cutorch/lib/THC/THCTensorTopk.cu
[THCTensorTopK.cu.txt](https://github.com/arrayfire/arrayfire/files/1146785/THCTensorTopK.cu.txt)

Cheers
W.

**Comment by WilliamTambellini** [comment]
some more ref:
https://arxiv.org/pdf/0906.0231.pdf
also shows a heap based parallel topk (fig.8).

**Comment by WilliamTambellini** [comment]
this slide also briefly presents a parallel k partial sort from slide 29:
http://on-demand.gputechconf.com/gtc/2009/presentations/1034-Multi-GPU-Recommendation-System.pdf

**Comment by WilliamTambellini** [comment]
Old discussion from the forum:
https://groups.google.com/forum/#!msg/arrayfire-users/oDtQcI7afZQ/AvN3xLycwWoJ;context-place=searchin/arrayfire-users/partial$20sort%7Csort:relevance

**Comment by WilliamTambellini** [comment]
Also found an "old" paper speaking about quickSelect:
"Parallel calculation of the median and order statistics on GPUs with application to robust regression"
[1104.2732.pdf](https://github.com/arrayfire/arrayfire/files/1149867/1104.2732.pdf)

**Comment by WilliamTambellini** [comment]
Found a cuda implementation of a "kth element" based on parallel median finder/bubblesort:
https://github.com/mindbane/Parallel-Median-of-Medians-CUDA

**Comment by WilliamTambellini** [comment]
Perhaps the best references I ve found: 
Google "Grinnell GPU K-Selection" implementation : 
https://code.google.com/archive/p/ggks/
The paper:
http://www.math.grin.edu/~blanchaj/Research/ABGS_KSelection.pdf
@umar456  does it sounds good ?

**Comment by WilliamTambellini** [comment]
Another very good paper: "Randomized Selection on the GPU":
http://www.lanl.gov/orgs/adtsc/publications/science_highlights_2012/docs/D_Monroe.pdf
"Both keys and values are extracted. The original list is untouched." which is exactly what I would like.
[D_Monroe - Randomized selection on the GPU.pdf](https://github.com/arrayfire/arrayfire/files/1157874/D_Monroe.-.Randomized.selection.on.the.GPU.pdf)

**Comment by QuantScientist** [comment]
very interesting. What is your use case for this?

**Comment by WilliamTambellini** [comment]
Hi @QuantScientist  My use case is the regular NN post softmax best candidates selection.

News: an openCL/ViennaCL implementation of topk via radix select:
https://github.com/vpa1977/smalltests/blob/master/radix_select.hpp
line 296:
/*
Select N smallest values from basic_type vector
*/
template<typename basic_type>
struct radix_select
{
....
radix_select(int size, const viennacl::context& ctx) {
}
}

**Comment by WilliamTambellini** [comment]
Facebook implementation over cuda:
https://github.com/facebook/fbcuda/blob/master/TopKElements.cuh

**Comment by WilliamTambellini** [comment]
If I resume correctly, the options are:
- quickSelect
- radixSelect
- bucketSelect
- randomizedSelect
To be continued...

**Comment by WilliamTambellini** [comment]
Funny: both Tensorflow and Theano are also working on gpu topk: 
https://github.com/tensorflow/tensorflow/issues/5719
https://github.com/Theano/Theano/pull/5959
Looks like they are more into radixSelect:
https://github.com/Theano/Theano/pull/5959/commits/ae212a7680d860879b7abb6f4954b72c8b2e5f15#diff-82c09ae646dfbba2e8ae35c42c61486b

**Comment by WilliamTambellini** [comment]
Naive cuda implementation of radixSelect from stackoverflow : unknown performance:  
[radixSelect.cu.txt](https://github.com/arrayfire/arrayfire/files/1183858/radixSelect.cu.txt)
Makefile:
radixSelect: radixSelect.cu
	nvcc -gencode=arch=compute_60,code=sm_60 -gencode=arch=compute_60,code=compute_60 --compiler-options -fno-strict-aliasing -I/usr/local/cuda/include -I/usr/local/cuda-8.0/samples/common/inc -DUNIX radixSelect.cu -o radixSelect -O3

**Comment by pavanky** [comment]
@umar456 Once this is done, don't forget to modify `nearestNeighbour` to use topk internally.

**Comment by WilliamTambellini** [comment]
Found an 'old' article but still relevant about radix select strategies:
"Radix selection algorithm for the kth order statistic" 
[Radix selection algo for the kth order stat - a4.pdf](https://github.com/arrayfire/arrayfire/files/1230254/Radix.selection.algo.for.the.kth.order.stat.-.a4.pdf)
@umar456  would you have any early ProofOfConcept results from your experiments ?
Cheers

**Comment by WilliamTambellini** [comment]
Interesting: few days ago, someone has implemented (at least in cuda) and merged into TensorFlow a  heap-based topk: 
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/topk_op_gpu.cu.cc
The license is Apache so I suppose it should be allowed to adapt it for AF.
Someone reports tedious performance compared to CPU but he is using very small input arrays.
TBC

**Comment by umar456** [comment]
Implemented via #2061
